src/fstar/FStar.CheckedFiles.fs

```diff
@@ -0,0 +1,335 @@
+﻿(*
+   Copyright 2008-2018 Microsoft Research
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*)
+#light "off"
+
+module FStar.CheckedFiles
+open FStar.ST
+open FStar.Exn
+open FStar.All
+open FStar
+open FStar.Errors
+open FStar.Util
+open FStar.Getopt
+open FStar.Syntax.Syntax
+open FStar.Extraction.ML.UEnv
+open FStar.TypeChecker.Env
+open FStar.Syntax.DsEnv
+
+(* Module abbreviations for the universal type-checker  *)
+module Syntax  = FStar.Syntax.Syntax
+module TcEnv   = FStar.TypeChecker.Env
+module SMT     = FStar.SMTEncoding.Solver
+module BU      = FStar.Util
+module Dep     = FStar.Parser.Dep
+
+
+(*
+ * We write this version number to the cache files, and
+ * detect when loading the cache that the version number is same
+ *)
+let cache_version_number = 9
+
+type tc_result = {
+  checked_module: Syntax.modul; //persisted
+  mii:module_inclusion_info; //persisted
+  smt_decls:(FStar.SMTEncoding.Term.decls_t *  //list of smt decls and fvbs for the module
+             list<FStar.SMTEncoding.Env.fvar_binding>); //persisted
+
+  tc_time:int;
+  extraction_time:int
+}
+
+(*
+ * Abbreviation for what we store in the checked files
+ *)
+type checked_file_entry =
+  int *                      //cache version number
+  list<(string * string)> *  //digest of direct dependencies
+  string *                   //digest of just this file
+  Parser.Dep.parsing_data *  //parsing data for this file
+  tc_result                  //typechecking result, including the smt encoding
+
+(*
+ * Local cache for checked files contents
+ * Note that a checked file could have valid parsing data but stale tc data
+ *)
+
+(*
+ * Cache files could be loaded in two steps
+ * 
+ * Initially the dependence analysis is just interested in the parsing data
+ *   and till that point we don't have the dependences sorted out, because of
+ *   which we can't check the validity of tc data (since we need to check hashes
+ *   of direct dependences etc.)
+ *
+ * So in this step, we read the checked file and mark the validity if tc data as Unknown
+ *
+ * Later on, we have figured the complete dependence graph, and want to load
+ *   the tc data
+ *
+ *  At that point, the cache is updated to either Valid or Invalid w.r.t. the tc data
+ *)
+type tc_result_t =
+  | Unknown of (list<(string * string)>) * string * tc_result  //deps string, digest of this file, tc_result
```

Comment:
The first two fields here correspond to

```
list<(string * string)> *  //digest of direct dependencies
  string *                   //digest of just this file
```

in the checked_file_entry, right?

I wonder if it is worth defining

`type deps_and_digest = list<(string * string)> * string`

and using that type more uniformly.

That would require a cache number increment if we change it in checked_file_entry. So, maybe we should wait to bundle this cosmetic change with some other future change that really requires a cache file increment